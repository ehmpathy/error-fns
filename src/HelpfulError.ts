import { omit } from 'type-fns';

export type HelpfulErrorMetadata = Record<string, any> & { cause?: Error };

/**
 * HelpfulError errors are used to add information that helps the future observer of the error understand whats going on
 */
export class HelpfulError extends Error {
  constructor(message: string, metadata?: HelpfulErrorMetadata) {
    const metadataWithoutCause = metadata
      ? omit(metadata, ['cause'])
      : metadata;
    const fullMessage = `${message}${
      metadataWithoutCause && Object.keys(metadataWithoutCause).length
        ? `\n\n${JSON.stringify(metadataWithoutCause)}`
        : ''
    }`;
    super(fullMessage, metadata?.cause ? { cause: metadata.cause } : undefined);
  }

  /**
   * a utility to throw an error of this class, for convenience
   *
   * e.g.,
   * ```ts
   * const phone = customer.phone ?? HelpfulError.throw('expected a phone');
   * ```
   */
  public static throw<T extends typeof HelpfulError>(
    this: T, // https://stackoverflow.com/a/51749145/3068233
    message: string,
    metadata?: HelpfulErrorMetadata,
  ): never {
    // eslint-disable-next-line @typescript-eslint/no-throw-literal
    throw new this(message, metadata) as InstanceType<T>;
  }
}
