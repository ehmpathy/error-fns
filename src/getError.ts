/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { isAFunction, isAPromise } from 'type-fns';

export class NoErrorThrownError extends Error {}

/**
 * a method which returns the error thrown from executing the logic passed into it
 *
 * usecases
 * - testing that a some logic throws an error
 *
 * note
 * - uses type overloading, so if you pass in a promise or async function, you'll be given a promise; otherwise, sync
 */
export function getError<T = Error>(
  from: () => Promise<unknown>,
): Promise<T | NoErrorThrownError>;
export function getError<T = Error>(
  from: Promise<unknown>,
): Promise<T | NoErrorThrownError>;
export function getError<T = Error>(
  from: () => unknown,
): T | NoErrorThrownError;
export function getError<T = Error>(
  from: Promise<unknown> | (() => Promise<unknown>) | (() => unknown),
) {
  // run everything in a try catch, to handle the case where the function throws an error immediately (does not return a promise)
  try {
    // execute the function to get the result, if its a function
    const result = isAFunction(from) ? from() : from;

    // handle the result = promise case
    if (isAPromise(result))
      return result
        .then(() => {
          throw new NoErrorThrownError();
        })
        .catch((error) => error as T | NoErrorThrownError);

    // handle the result = nonpromise case
    throw new NoErrorThrownError(); // should have never reached here, as it should have thrown when we first called it
  } catch (error: unknown) {
    return error as T;
  }
}
