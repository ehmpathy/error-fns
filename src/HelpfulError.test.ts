import { HelpfulError } from './HelpfulError';
import { getError } from './getError';

describe('HelpfulError', () => {
  it('should produce a helpful, observable error message', () => {
    const error = new HelpfulError('the dogs were let out', {
      who: 'your mom', // ðŸ™„ðŸ˜‚
    });
    expect(error).toMatchSnapshot();
  });
  it('should be throwable in a ternary conveniently and precisely', () => {
    const error = getError(() => {
      // this case should not throw
      const customerOne: { phone: string | null } = {
        phone: 'yes',
      };
      const phoneOne =
        customerOne.phone ?? HelpfulError.throw('phone one not found!');

      // but this case should throw
      const customerTwo: { phone: string | null } = {
        phone: null,
      };
      const phoneTwo =
        customerTwo.phone ?? HelpfulError.throw('phone two not found!');
    });
    expect(error).toBeInstanceOf(HelpfulError);
    expect(error.message).toEqual('phone two not found!');
  });
  it('should be possible to extend the call stack of an error via cause', () => {
    const errorOriginal = new Error('some original error');
    const errorHelpful = new HelpfulError('some helpful error', {
      cause: errorOriginal,
    });
    expect(errorHelpful).toMatchSnapshot();
    expect(errorHelpful.cause).toBeDefined();
    expect(errorHelpful.cause).toMatchSnapshot();
  });
});
