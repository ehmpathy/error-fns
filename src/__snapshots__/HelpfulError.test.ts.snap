// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpfulError should be possible to extend the call stack of an error via cause 1`] = `[Error: some helpful error]`;

exports[`HelpfulError should be possible to extend the call stack of an error via cause 2`] = `[Error: some original error]`;

exports[`HelpfulError should produce a helpful, observable error message 1`] = `
[Error: the dogs were let out

{"who":"your mom"}]
`;
